src/net/udp.rs

```diff
@@ -40,41 +50,50 @@ impl UdpSocket {
             .map_err(|io| NetworkErrorKind::IOError { inner: io })?;
 
         if len > 0 {
-            let packet = self.recv_buffer[..len].to_owned();
+            let packet = &self.recv_buffer[..len];
+
+            let connection = self.connections.get_connection_or_insert(&addr)?;
+            let mut lock = connection
+                .write()
+                .map_err(|error| NetworkError::poisoned_connection_error(error.description()))?;
+
+            return lock.process_incoming(&packet)
 
-            self.packet_processor
-                .process_data(packet, addr, &mut self.state)
         } else {
             Err(NetworkErrorKind::ReceivedDataToShort)?
         }
     }
 
     /// Sends data on the socket to the given address. On success, returns the number of bytes written.
     pub fn send(&mut self, packet: Packet) -> NetworkResult<usize> {
-        let (addr, mut packet_data) = self.state.pre_process_packet(packet, &self.config)?;
+        let connection = self.connections.get_connection_or_insert(&packet.addr())?;
+        let mut lock = connection
+            .write()
+            .map_err(|error| NetworkError::poisoned_connection_error(error.description()))?;
+
+        let mut packet_data = lock.process_outgoing(packet.payload(), packet.delivery_method())?;
+
         let mut bytes_sent = 0;
+
         if let Some(link_conditioner) = &self.link_conditioner {
             if link_conditioner.should_send() {
                 for payload in packet_data.parts() {
-                    bytes_sent += self.socket.send_to(&payload, addr).map_err(|io| {
+                    bytes_sent += self.socket.send_to(&payload, packet.addr()).map_err(|io| {
                         NetworkError::from(NetworkErrorKind::IOError { inner: io })
                     })?;
                 }
             }
         } else {
             for payload in packet_data.parts() {
                 bytes_sent += self
-                    .socket
-                    .send_to(&payload, addr)
-                    .map_err(|io| NetworkError::from(NetworkErrorKind::IOError { inner: io }))?;
+                .socket
+                .send_to(payload, packet.addr())
+                .map_err( |io | NetworkError::from(NetworkErrorKind::IOError { inner: io })) ?;
```

Comment:
spaces 😄

Reply:
Will be fixed with cargo fmt I'll perform before merging.